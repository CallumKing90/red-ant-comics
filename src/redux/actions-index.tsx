import { actions as applicationStateActions } from "./reducers/applicationState/applicationState";

export default {
  ...applicationStateActions,
};
