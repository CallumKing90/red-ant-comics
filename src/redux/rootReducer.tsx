import { combineReducers } from "@reduxjs/toolkit";
import applicationState from "./reducers/applicationState/applicationState";
import { marvelApi } from "./services/marvel/marvelApi";

export const rootReducer = combineReducers({
  applicationState,
  [marvelApi.reducerPath]: marvelApi.reducer,
});

export default rootReducer;
