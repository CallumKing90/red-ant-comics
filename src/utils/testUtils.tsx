import React from "react";
import { createStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";

export const renderWithRedux = (
  ui: JSX.Element,
  store = createStore(() => ({}))
) => ({
  ...render(<Provider store={store}>{ui}</Provider>),
  store,
});
