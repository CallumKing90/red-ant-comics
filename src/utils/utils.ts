interface Favourite {
  title: string;
  id: string;
}

export const comicIdInArray = (
  favouritesList: Favourite[],
  comicId: string
) => {
  for (const favourite of favouritesList) {
    if (favourite.id === comicId) return true;
  }

  return false;
};
