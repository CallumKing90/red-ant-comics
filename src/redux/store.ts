import { configureStore } from "@reduxjs/toolkit";
import rootReducer from "./rootReducer";
import { marvelApi } from "./services/marvel/marvelApi";

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(marvelApi.middleware),
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
