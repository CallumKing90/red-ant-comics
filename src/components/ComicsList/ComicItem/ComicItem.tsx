import React from "react";
import { useAppSelector, useAppDispatch } from "../../../utils/hooks";
import { Button } from "../../Button/Button";
import { Heading } from "../../Heading/Heading";
import { ComicImage } from "../ComicImage/ComicImage";
import { comicIdInArray } from "../../../utils/utils";
import { getFavourites } from "../../../redux/selectors/applicationState/applicationStateSelector";
import actions from "../../../redux/actions-index";
import Comic from "../../../ts/types/Comic";

interface comicProps {
  comic: Comic;
}

export function ComicItem({ comic }: comicProps): JSX.Element {
  const dispatch = useAppDispatch();
  const favourites = useAppSelector(getFavourites);
  const { title, id, thumbnail } = comic;

  const isAddedToFavourites = comicIdInArray(favourites, id);

  const addToFavourites = () => {
    dispatch(actions.addToFavourites({ title, id }));
  };

  const removeFromFavourites = () => {
    dispatch(actions.removeFromFavourites(id));
  };

  return (
    <li className="comic-item" id={id} data-testid="comic-item">
      <div className="comic-card">
        <ComicImage
          title={title}
          variation="standard_fantastic"
          thumbnail={thumbnail}
        />
        <Heading variation="h2">{title}</Heading>
        {isAddedToFavourites ? (
          <Button className="button js-add" onClick={removeFromFavourites}>
            Remove from favourites
          </Button>
        ) : (
          <Button className="button js-add" onClick={addToFavourites}>
            Add to favourites
          </Button>
        )}
      </div>
    </li>
  );
}
