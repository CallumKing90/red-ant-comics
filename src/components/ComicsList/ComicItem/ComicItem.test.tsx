import React from "react";
import { ComicItem } from "./ComicItem";
import "@testing-library/jest-dom";
import { renderWithRedux } from "../../../utils/testUtils";
import { store } from "../../../redux/store";
import { createStore } from "@reduxjs/toolkit";
import { fireEvent } from "@testing-library/react";

const testComic = {
  title: "TestComic",
  id: "1234",
  thumbnail: {
    path: "pathToThumbnail",
    extension: "jpg",
  },
};

describe("ComicItem", () => {
  it("should render the ComicItem component", () => {
    const { getByTestId } = renderWithRedux(
      <ComicItem comic={testComic} />,
      store
    );
    const comicItem = getByTestId("comic-item");
    expect(comicItem).toBeInTheDocument();
    expect(comicItem).toMatchSnapshot();
  });

  it("should show 'Add to Favourites' button when id not in favourites array", () => {
    const { getByText } = renderWithRedux(
      <ComicItem comic={testComic} />,
      store
    );
    let comicItemButton = getByText("Add to favourites");
    expect(comicItemButton).toBeInTheDocument();
    expect(comicItemButton).toMatchSnapshot();

    fireEvent.click(comicItemButton);

    comicItemButton = getByText("Remove from favourites");
    expect(comicItemButton).toBeInTheDocument();
    expect(comicItemButton).toMatchSnapshot();
  });

  it("should show 'Remove from Favourites' button when id not in favourites array", () => {
    const { getByText } = renderWithRedux(
      <ComicItem comic={testComic} />,
      store
    );

    let comicItemButton = getByText("Remove from favourites");
    expect(comicItemButton).toBeInTheDocument();
    expect(comicItemButton).toMatchSnapshot();

    fireEvent.click(comicItemButton);

    comicItemButton = getByText("Add to favourites");
    expect(comicItemButton).toBeInTheDocument();
    expect(comicItemButton).toMatchSnapshot();
  });
});
