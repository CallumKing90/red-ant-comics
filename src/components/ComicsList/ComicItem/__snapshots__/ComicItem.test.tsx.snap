// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComicItem should render the ComicItem component 1`] = `
<li
  class="comic-item"
  data-testid="comic-item"
  id="1234"
>
  <div
    class="comic-card"
  >
    <img
      alt="TestComic"
      data-testid="comic-image"
      loading="lazy"
      role="img"
      src="pathToThumbnail/standard_fantastic.jpg"
    />
    <h2>
      TestComic
    </h2>
    <button
      class="button js-add"
      role="button"
    >
      Add to favourites
    </button>
  </div>
</li>
`;

exports[`ComicItem should show 'Add to Favourites' button when id not in favourites array 1`] = `
<button
  class="button js-add"
  role="button"
>
  Add to favourites
</button>
`;

exports[`ComicItem should show 'Add to Favourites' button when id not in favourites array 2`] = `
<button
  class="button js-add"
  role="button"
>
  Remove from favourites
</button>
`;

exports[`ComicItem should show 'Remove from Favourites' button when id not in favourites array 1`] = `
<button
  class="button js-add"
  role="button"
>
  Remove from favourites
</button>
`;

exports[`ComicItem should show 'Remove from Favourites' button when id not in favourites array 2`] = `
<button
  class="button js-add"
  role="button"
>
  Add to favourites
</button>
`;
